File: Union Find/Quick Union/QuickUnion.c
//Quick Find
#include<stdio.h>
#include<stdlib.h>
int *id = NULL;
int n = 0;
int root(int i){
    while(i != id[i]){
        i = id[i];
    }
    return i;
}
int connected(int p,int q){
    if(root(p) == root(q))
    {
        return 1;
    }
    return 0;
}

void Union(int p,int q){
    int i = root(p);
    int j = root(q);
    id[i] = j;
}



int main(){
    
    printf("Enter the value of n\n");
    scanf("%d",&n);
    id = (int*)malloc(sizeof(int)*n);
    for(int i = 0;i<n;i++){
        id[i] = i;
    }
    Union(1,2);
    Union(2,3);
    
    if(connected(1,4))
        printf("True");
    else
        printf("False");
    return 0;
}

File: Union Find/Quick Union/QuickUnion.java
package com.company;

public class QuickUnionUF {
    private int id[];
    public QuickUnionUF(int n){
        id= new int[n];
        for(int i= 0;i<n;i++){
            id[i] = i;
        }
    }
    public int root(int r){
        while(r != id[r]){
            r = id[r];
        }
        return r;
    }
    public boolean connected(int p,int q){
        return root(p) == root(q);
    }
    public void union(int p,int q){
        int i = root(p);
        int j = root(q);
        id[i] = j;
    }
    public static void main(String args[]){
        QuickUnionUF a = new QuickUnionUF(5);
        a.union(1,2);
        a.union(4,3);
        a.union(3,2);
        System.out.println(a.connected(1,4));
    }
}
//Defects of Quick Union is Trees can get to tall, therefore find operation will consume more time.
             
